src/tpm2/RuntimeAlgorithm.c

```diff
@@ -0,0 +1,506 @@
+/*										*/
+/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS		*/
+/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT		*/
+/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR	*/
+/* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT		*/
+/* HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,	*/
+/* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT		*/
+/* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,	*/
+/* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY	*/
+/* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT		*/
+/* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE	*/
+/* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.		*/
+/*										*/
+/********************************************************************************/
+
+#define _GNU_SOURCE
+#include <assert.h>
+#include <string.h>
+
+#include "Tpm.h"
+#include "NVMarshal.h"
+#include "GpMacros.h"
+#include "tpm_library_intern.h"
+
+#define ALGO_SEPARATOR_C ','
+#define ALGO_SEPARATOR_STR ","
+
+struct KeySizes {
+    BOOL enabled;
+    UINT16 size;
+};
+
+static const struct KeySizes s_KeySizesAES[] = {
+    { .enabled = AES_128, .size = 128 },
+    { .enabled = AES_192, .size = 192 },
+    { .enabled = AES_256, .size = 256 },
+    { .enabled = false  , .size = 0 },
+};
+static const struct KeySizes s_KeySizesSM4[] = {
+    { .enabled = SM4_128, .size = 128 },
+    { .enabled = false  , .size = 0 },
+};
+static const struct KeySizes s_KeySizesCamellia[] = {
+    { .enabled = CAMELLIA_128, .size = 128 },
+    { .enabled = CAMELLIA_192, .size = 192 },
+    { .enabled = CAMELLIA_256, .size = 256 },
+    { .enabled = false       , .size = 0 },
+};
+static const struct KeySizes s_KeySizesTDES[] = {
+    { .enabled = TDES_128, .size = 128 },
+    { .enabled = TDES_192, .size = 192 },
+    { .enabled = false   , .size = 0 },
+};
+static const struct KeySizes s_KeySizesRSA[] = {
+    { .enabled = RSA_1024, .size = 1024 },
+    { .enabled = RSA_2048, .size = 2048 },
+    { .enabled = RSA_3072, .size = 3072 },
+    { .enabled = RSA_4096, .size = 4096 },
+    { .enabled = false   , .size = 0 },
+};
+static const struct KeySizes s_KeySizesECC[] = {
+    { .enabled = ECC_NIST_P192, .size = 192 },
+    { .enabled = ECC_NIST_P224, .size = 224 },
+    { .enabled = ECC_NIST_P256, .size = 256 },
+    { .enabled = ECC_BN_P256  , .size = 256 },
+    { .enabled = ECC_SM2_P256 , .size = 256 },
+    { .enabled = ECC_NIST_P384, .size = 384 },
+    { .enabled = ECC_NIST_P521, .size = 521 },
+    { .enabled = ECC_BN_P638  , .size = 638 },
+    { .enabled = false        , .size = 0 },
+};
+
+static const struct {
+    const char   *name;
+    union {
+	const struct KeySizes *keySizes;
+    } u;
+    BOOL          canBeDisabled;
+} s_AlgorithmProperties[NUM_ENTRIES_ALGORITHM_PROPERTIES] = {
```

Comment:
Add a column `stateFormatLevel` that indicates which stateFormatLevel is required to write this algorithm?